### ISSUE TITLE
Unknown 'SDES' flag encountered: 'pad', OSRTP-offer not generating a:crypto lines

### ISSUE THREAD

**Issue by amessina** [sipwise/rtpengine#976]
Testing the latest master at e741a7434eb14aefa0073b9b67a6753dae3a1bc3 with updates for `SDES-pad` and `OSRTP-offer` and kamailio-5.3.3.  I've also tried updating kamailio's rtpengine module to the latest updates in the kamailio master branch in case there are kamailio parsing updates.

kamailio `rtpengine_manage()` flags:
```
via-branch=extra replace-origin replace-session-connection strict-source SDES-pad direction=pub direction=priv address-family=IP6 record-call OSRTP-offer AVP ICE=remove
```

Replacing `pad-crypto` with `SDES-pad` on a call where the endpoint requires it:
```
WARNING: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Unknown 'SDES' flag encountered: 'pad'
```

Using `OSRTP-offer` where the incoming offer has no `a=crypto` lines, expecting rtpengine to insert `a=crypto` lines to the recipient, none are inserted:

```
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Response dump for 'offer' to 127.0.0.1:38453: { "sdp": "v=0
o=root 340457006 340457006 IN IP6 2603:300a:134:50e0::3
s=session
c=IN IP6 2603:300a:134:50e0::3
t=0 0
m=audio 30000 RTP/AVP 0 8 18 101
a=silenceSupp:off - - - -
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:18 G729/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:18 annexb=no
a=fmtp:101 0-16
a=sendrecv
a=rtcp:30001
a=ptime:20
", "result": "ok" }
```

Full output:

```
INFO: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Received command 'offer' from 127.0.0.1:38453
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Dump for 'offer' from 127.0.0.1:38453: { "supports": [ "load limit" ], "sdp": "v=0
o=root 340457006 340457006 IN IP4 204.11.192.134
s=session
c=IN IP4 204.11.192.134
t=0 0
m=audio 17314 RTP/AVP 0 8 18 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:18 G729/8000
a=fmtp:18 annexb=no
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=silenceSupp:off - - - -
a=ptime:20
a=sendrecv
", "address-family": "IP6", "ICE": "remove", "direction": [ "pub", "priv" ], "flags": [ "strict-so ...
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: ... urce", "record-call", "OSRTP-offer" ], "replace": [ "origin", "session-connection" ], "transport-protocol": "RTP/AVP", "SDES": [ "pad" ], "call-id": "6ab42e533853d8e764ce475805db8826@204.11.192.134", "via-branch": "z9hG4bK3e9e54290", "received-from": [ "IP4", "204.11.192.134" ], "from-tag": "as1b714123", "command": "offer" }
WARNING: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Unknown 'SDES' flag encountered: 'pad'
NOTICE: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Creating new call
NOTICE: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Turning on call recording.
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: kernel call idx is 0
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Default sink codec is PCMU/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: 0 DTMF sink entries
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Creating codec handler for PCMU/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Sink supports codec PCMU/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Using passthrough handler for PCMU/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Creating codec handler for PCMA/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Sink supports codec PCMA/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Using passthrough handler for PCMA/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Creating codec handler for G729/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Sink supports codec G729/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Using passthrough handler for G729/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Creating codec handler for telephone-event/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Sink supports codec telephone-event/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Using passthrough handler for telephone-event/8000
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: creating send_timer
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: creating send_timer
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: creating send_timer
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: creating send_timer
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: kernel stream idx is 0
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: kernel stream idx is 1
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: set FILLED flag for stream 204.11.192.134:17314
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: set FILLED flag for stream 204.11.192.134:17315
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: kernel stream idx is 2
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: kernel stream idx is 3
INFO: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Replying to 'offer' from 127.0.0.1:38453 (elapsed time 0.020635 sec)
DEBUG: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Response dump for 'offer' to 127.0.0.1:38453: { "sdp": "v=0
o=root 340457006 340457006 IN IP6 2603:300a:134:50e0::3
s=session
c=IN IP6 2603:300a:134:50e0::3
t=0 0
m=audio 30000 RTP/AVP 0 8 18 101
a=silenceSupp:off - - - -
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:18 G729/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:18 annexb=no
a=fmtp:101 0-16
a=sendrecv
a=rtcp:30001
a=ptime:20
", "result": "ok" }
```

**Comment by rfuchs** [comment]
The lack of OSRTP is due to your use of transport-protocol=RTP/AVP, which overrides the OSRTP stuff. It's supposed to work, so I'll push a fix for that next week.

**Comment by amessina** [comment]
Thank you @rfuchs -- I saw that and Kamailio is only sending `via-branch=extra replace-origin replace-session-connection strict-source SDES-pad direction=pub direction=priv address-family=IP6 record-call OSRTP-offer AVP ICE=remove` with `rtpengine_manage()` so I'm guessing it may be the Kamailio module that needs updating.  Have a good evening and thanks for your always attentive comments here.

**Comment by rfuchs** [comment]
It's the `AVP` part that makes it put the protocol in there.

**Comment by amessina** [comment]
Awww crap.  Ok--lemme try without that...  Would that also affect the `WARNING: [6ab42e533853d8e764ce475805db8826@204.11.192.134]: Unknown 'SDES' flag encountered: 'pad'`?

**Comment by rfuchs** [comment]
No, the fix for that is already in: 5caf47f

**Comment by amessina** [comment]
For reference, I had been doing a sort of poor man's version by being more specific in my Kamailio script with regard to RTP vs SRTP and AVP vs AVPF -- working toward a universal rtpengine route block that would handle essentially from anything to anything.  I'll need to rethink the Kamailio script part now that your OSRTP options have landed.

```
# RTP, SRTP or DTLS transport (offer SRTP by default)
if($avp(rp)=~"^WS") {
        # SDES-off needs to happen even on re-INVITE (in case of adding or removing video)
        # Have to move this out of the "first offer only" block...)
        $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " DTLS DTLS=passive SDES-off";
} else if(sdp_get_line_startswith("$avp(offerrer_crypto)", "a=crypto")) {
        if(isbflagset(FLB_RTPENGINE_RTP)) {
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " RTP";
        } else {
                # Default to SRTP
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " SRTP";
        }
} else {
        if(isbflagset(FLB_RTPENGINE_RTP)) {
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " RTP";
        } else {
                # Default to SRTP
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " SRTP";
                # Try OSRTP (think we need updates in kamailio's rtpengine module first)
                #$xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " OSRTP-offer";
        }
}

# AVP or AVPF feedback protocol (offer AVP by default)
if(sdp_get_line_startswith("$avp(caller_rtcpfb)", "a=rtcp-fb")) {
        if(isbflagset(FLB_RTPENGINE_AVPF) || $avp(rp)=~"^WS") {
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " AVPF";
        } else {
                # Default to AVP
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " AVP";
        }
} else {
        if(isbflagset(FLB_RTPENGINE_AVPF) || $avp(rp)=~"^WS") {
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " AVPF";
        } else {
                # Default to AVP
                $xavp(ro=>$T_branch_idx)=$xavp(ro=>$T_branch_idx) + " AVP";
        }
}
```

**Comment by amessina** [comment]
> No, the fix for that is already in: [5caf47f](https://github.com/sipwise/rtpengine/commit/5caf47f51f5425e851948fea738f9291126442d0)

Thank you!
